runtime/core.go

```diff
@@ -936,6 +959,83 @@ func compareDefault(f *Frame, v, w *Object) int {
 	return 1
 }
 
+// tryRichCompareBool tries a rich comparison with the given comparison op and
+// returns a bool indicating if the relation is true. It closely resembles the
+// behavior of CPython's try_rich_compare_bool in object.c.
+func tryRichCompareBool(f *Frame, op compareOp, v, w *Object) (bool, *BaseException) {
+	r, raised := compareRich(f, op, v, w)
+	if raised != nil {
+		return false, raised
+	}
+	if r == NotImplemented {
+		return false, nil
+	}
+	br, raised := IsTrue(f, r)
+	if raised != nil {
+		return false, raised
+	}
+	return br, raised
+}
+
+// halfCompare tries a comparison with the __cmp__ slot, ensures the result
+// is an integer, and returns it. It closely resembles the behavior of CPython's
+// half_compare in typeobject.c.
+func halfCompare(f *Frame, v, w *Object) (*Object, *BaseException) {
```

Comment:
In a subsequent PR, would you mind changing the return value to (*Int, *BaseException)? You pass the result to intNeg() later on in the code and it would be easier to understand if it was an *Int.